Mathlib/Analysis/NormedSpace/ProdLp.lean

```diff
@@ -0,0 +1,743 @@
+          apply ENNReal.rpow_le_rpow _ (by positivity)
+          simp only [self_le_add_right]
+    · calc
+        edist x.snd y.snd ≤ (edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) := by
+          simp only [← ENNReal.rpow_mul, cancel, ENNReal.rpow_one, le_refl]
+        _ ≤ (edist x.fst y.fst ^ p.toReal + edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) := by
+          apply ENNReal.rpow_le_rpow _ (by positivity)
+          simp only [self_le_add_left]
+
+example (a : ENNReal) : a + a = 2*a := by exact Eq.symm (two_mul a)
+
+theorem prod_antilipschitzWith_equiv_aux [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
+    AntilipschitzWith ((2 : ℝ≥0) ^ (1 / p).toReal) (WithLp.equiv p (α × β)) := by
+  intro x y
+  rcases p.dichotomy with (rfl | h)
+  · simp [edist]
+  · have pos : 0 < p.toReal := by positivity
+    have nonneg : 0 ≤ 1 / p.toReal := by positivity
+    have cancel : p.toReal * (1 / p.toReal) = 1 := mul_div_cancel' 1 (ne_of_gt pos)
+    rw [prod_edist_eq_add pos, ENNReal.toReal_div 1 p]
+    simp only [edist, ← one_div, ENNReal.one_toReal]
+    calc
+      (edist x.fst y.fst ^ p.toReal + edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) ≤
+          (edist (WithLp.equiv p _ x) (WithLp.equiv p _ y) ^ p.toReal +
+          edist (WithLp.equiv p _ x) (WithLp.equiv p _ y) ^ p.toReal) ^ (1 / p.toReal) := by
+        refine ENNReal.rpow_le_rpow (add_le_add ?_ ?_) nonneg
+        · refine ENNReal.rpow_le_rpow ?_ (le_of_lt pos)
+          simp [edist]
+        · refine ENNReal.rpow_le_rpow ?_ (le_of_lt pos)
+          simp [edist]
+      _ = (2 : ℝ≥0) ^ (1 / p.toReal) * edist (WithLp.equiv p _ x) (WithLp.equiv p _ y) := by
+        simp only [← two_mul, ENNReal.mul_rpow_of_nonneg _ _ nonneg,
+          ← ENNReal.rpow_mul, cancel, ENNReal.rpow_one, ← ENNReal.coe_rpow_of_nonneg _ nonneg,
+          coe_ofNat]
+
+theorem prod_aux_uniformity_eq [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
+    𝓤 (WithLp p (α × β)) = 𝓤[instUniformSpaceProd] := by
+  have A : UniformInducing (WithLp.equiv p (α × β)) :=
+    (prod_antilipschitzWith_equiv_aux p α β).uniformInducing
+      (prod_lipschitzWith_equiv_aux p α β).uniformContinuous
+  have : (fun x : WithLp p (α × β) × WithLp p (α × β) =>
+    ((WithLp.equiv p (α × β)) x.fst, (WithLp.equiv p (α × β)) x.snd)) = id :=
+    by ext i <;> rfl
+  rw [← A.comap_uniformity, this, comap_id]
+
+theorem prod_aux_cobounded_eq [PseudoMetricSpace α] [PseudoMetricSpace β] :
+    cobounded (WithLp p (α × β)) = @cobounded _ Prod.instBornology :=
+  calc
+    cobounded (WithLp p (α × β)) = comap (WithLp.equiv p (α × β)) (cobounded _) :=
+      le_antisymm (prod_antilipschitzWith_equiv_aux p α β).tendsto_cobounded.le_comap
+        (prod_lipschitzWith_equiv_aux p α β).comap_cobounded_le
+    _ = _ := comap_id
+
+end Aux
+
+/-! ### Instances on `L^p` products -/
+
+
+instance instUniformSpace [UniformSpace α] [UniformSpace β] : UniformSpace (WithLp p (α × β)) :=
+  instUniformSpaceProd
```

Comment:
Maybe we should do this at the same time for the pi and the prod version? I've played around doing it for pi and there a few errors here and there, so I would suggest we do this after this PR. Similarly I start to think that all these instances should be in `WithLp`.

Reply:
Yes, I think putting the topology and uniform space (and bornology probably) in `WithLp` probably makes sense. I think the whole point of `WithLp p α` is that it's `α`, except you replace the norm structure with Lp, and leave everything else untouched.

I'm okay if we leave the defeq abuse until a future PR. Yes, it does require a (small) refactor to get everything working again.

Reply:
Just to comment that I already considered doing this, but was unsure whether it would be meaningful on things like `Finsupp` and `DFinsupp`, for which we could presumably attach exactly the same norm.